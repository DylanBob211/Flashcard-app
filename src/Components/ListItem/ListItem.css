/* Card */
.listItem_container {
    border-radius: 10px;
    background: var(--bg-color);
    box-shadow: -9px 17px 21px -16px rgba(0,0,0,0.49);
}

/* Header */
.listItem_header_container {
    border-radius: 10px 10px 0 0;
    display: flex;
    align-items: center;
    background-color: var(--primary-color)
}

.listItem_title {
    flex-grow: 4;
    text-align: center;;
    color: #fff;

}

.listItem_iconbox{
    flex-grow: 1;
    align-items: center;
}

.listItem_icon--trashbin {
    height: 20px;
    fill: #fff;
    transition: .3s fill;
}

.listItem_icon--play {
    height: 20px;
    margin-left: .2em;
    fill: #fff;
    transition: .3s fill;
}

.listItem_icon--trashbin:hover {
    fill: var(--first-shade-clashing);
    cursor: pointer;
}

.listItem_icon--play:hover {
    fill: var(--second-shade-secondary);
    cursor: pointer;
}

/* Body */

.listItem_wordList {
    padding: .5em;
}



