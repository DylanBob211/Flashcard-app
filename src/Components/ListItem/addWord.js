import Unsplash from 'unsplash-js'

const unsplash = new Unsplash({
    applicationId: "66c8f4bbdcd250ba88db15412fb4cf728a9a8c4db261db2a356b22a1365aa20a"
});

const addWord = (word, setWords) => {
    if(!word) {
        console.warn('Inserisci una parola')
    } else {

        const newWord = {
            word: word,
            url: ''
        }

        unsplash.search.photos(word)
            .catch(e => {
                console.log(e)
                const msg = { 
                    error : e
                }
                return JSON.stringify(msg)
            })
            .then(res => res.json())
            .catch(e => console.log(e))
            .then(res => newWord.url = res.results[0].urls.small)
            .catch(res => newWord.url = 'No Img Available')
            .then(_ => setWords(state => ([newWord, ...state])))    
    
    }
         
        
}

export default addWord