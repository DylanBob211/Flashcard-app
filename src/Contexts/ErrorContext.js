import React, { useState, useContext } from 'react';

const ErrorContext = React.createContext(['', () => {}]);

export const useErrorContext = () => useContext(ErrorContext);

const ErrorProvider = ({ children }) => {
  const [error, setError] = useState('');
  return (
    <ErrorContext.Provider
      value={[
        error,
        setError,
      ]}
    >
      {children}
    </ErrorContext.Provider>
  );
};

export default ErrorProvider;
