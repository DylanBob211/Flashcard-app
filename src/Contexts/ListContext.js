import React, { useContext } from 'react';
import initialState from '../Components/initialState';
import useLists from '../Hooks/useLists';

const ListContext = React.createContext([initialState, () => {}]);

export const useListContext = () => useContext(ListContext);

const ListProvider = ({ children }) => {
  const { lists, setLists, addNewList } = useLists();
  return (
    <ListContext.Provider value={{ lists, setLists, addNewList }}>
      {children}
    </ListContext.Provider>
  );
};

export default ListProvider;
