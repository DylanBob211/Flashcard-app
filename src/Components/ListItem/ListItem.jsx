import React from 'react';
import WordItem from '../WordItem/WordItem';
import WordForm from '../WordForm/WordForm';
import './ListItem.css';
import TrashBinIcon from '../Icons/TrashBinIcon';
import PlayIcon from '../Icons/PlayIcon';


const ListItem = ({
  id, wordsArray, listName, deleteList, addWord, deleteWord
}) => {
  const listOfWords = wordsArray.map((word, index) => (
    <WordItem
      deleteWord={deleteWord}
      name={word.word}
      picUrls={word.url}
      key={index}
      id={id}
    />
  ));

  return (
    <div className="listItem_container">
      <ListHeader
        listName={listName}
        deleteList={deleteList}
      />
      <ul className="listItem_wordList">{ listOfWords }</ul>
      <WordForm
        addWord={addWord}
        id={id}
      />
    </div>
  );
};

const ListHeader = ({ listName, deleteList }) => (
  <div className="listItem_header_container">
    <h2 className="listItem_title">{ listName }</h2>
    <div className="listItem_iconbox">
      <TrashBinIcon function={deleteList} dependencies={listName} class="listItem_icon--trashbin" />
      <PlayIcon class="listItem_icon--play" />
    </div>
  </div>
);

export default ListItem;
