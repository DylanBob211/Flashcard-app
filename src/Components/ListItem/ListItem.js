import React, {useState, useEffect} from 'react'
import WordItem from '../WordItem';
import WordForm from '../WordForm'
import addWord from './addWord'
import './ListItem.css'
import TrashBinIcon from '../Icons/TrashBinIcon';
import PlayIcon from '../Icons/PlayIcon'


const ListItem = ({ wordsArray, listName, deleteList }) => {
    
    const [words, setWords] = useState(wordsArray)

    const [isExpanded, setExpanded] = useState(false) //TODO: finish
         
    const listOfWords = words.map((word, index) => (<WordItem name={word.word} picUrl={word.url} key={index}/>))

    return (<div className="listItem_container" >
                <ListHeader listName={ listName } deleteList={ deleteList }/>
                <ul className="listItem_wordList">{ listOfWords }</ul> 
                <WordForm addWord={ addWord } dependency ={ setWords }/>
            </div>)
        
    
}

const ListHeader = ({listName, deleteList }) => {
    
    return (
    <div className="listItem_header_container">
        <h2 className="listItem_title">{ listName }</h2>
        <div className="listItem_iconbox">
            <TrashBinIcon function={deleteList} dependencies={listName} class="listItem_icon--trashbin"/>
            <PlayIcon class="listItem_icon--play" />
        </div>
    </div>
    )
}

export default ListItem