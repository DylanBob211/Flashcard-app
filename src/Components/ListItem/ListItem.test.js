import React from 'react';
import { mount, shallow } from 'enzyme';
import ListItem from './ListItem';
import * as ListContext from '../../Contexts/ListContext';

describe('ListItem', () => {
  const useListContext = jest.spyOn(ListContext, 'useListContext');
  const addWord = jest.fn();
  const deleteWord = jest.fn();
  const addNewList = jest.fn();
  const deleteList = jest.fn();
  useListContext.mockImplementation(() => ({
    addNewList,
    deleteList,
    addWord,
    deleteWord,
  }));

  let mountedListItem;
  const listItem = (props, isShallow = false) => {
    const compJsx = (
      <ListItem
        {...props}
      />
    );
    if (isShallow) {
      return shallow(compJsx);
    }
    return mount(compJsx);
  };

  beforeEach(() => {
    mountedListItem = undefined;
    jest.clearAllMocks();
  });

  describe('when wordsArray prop empty', () => {
    const propsMock = {
      listId: 'listIdMock',
      wordsArray: [],
      listName: 'listNameMock',
      deleteWord,
      openFlashcard: jest.fn(),
      openExerciseWindow: jest.fn(),
    };

    beforeEach(() => {
      addWord.mockReturnValue(() => {});
      deleteWord.mockReturnValue(() => {});
      propsMock.openExerciseWindow.mockReturnValue(() => {});
      propsMock.openFlashcard.mockReturnValue(() => {});
      mountedListItem = listItem(propsMock);
    });

    it('matches the snapshot', () => {
      const snapshotComp = listItem(propsMock, true);
      expect(snapshotComp).toMatchSnapshot();
    });

    it('always renders a div container', () => {
      const divContainer = mountedListItem.find('[data-test="listItemContainer"]');
      expect(divContainer.exists()).toEqual(true);
    });

    it('always renders a ListHeader component', () => {
      const listHeader = mountedListItem.find('ListHeader');
      expect(listHeader.exists()).toEqual(true);
    });

    it('always renders a WordForm', () => {
      const wordForm = mountedListItem.find('WordForm');
      expect(wordForm.exists()).toEqual(true);
    });

    it('curries the openExerciseWindow prop with the listId', () => {
      const { openExerciseWindow, listId } = propsMock;
      expect(openExerciseWindow).toHaveBeenCalledWith(listId);
    });

    it('passes it the curried openExerciseWindow to the ListHeader component', () => {
      const listHeader = mountedListItem.find('ListHeader');
      const { openExerciseWindow, listId } = propsMock;
      const curried = openExerciseWindow(listId);
      expect(listHeader.props().openExerciseWindow).toEqual(curried);
    });

    it('curries `addWord` with `listId`', () => {
      const { listId } = propsMock;
      expect(addWord).toHaveBeenCalledWith(listId);
    });

    it('passes the curried `addWord` to the WordForm component', () => {
      const wordForm = mountedListItem.find('WordForm');
      const { listId } = propsMock;
      const curried = addWord(listId);
      expect(wordForm.props().addWord).toEqual(curried);
    });
  });

  describe('when wordsArray prop has elements', () => {
    const wordsArray = [
      { word: 'hello', url: ['helloUrl'] },
      { word: 'word', url: ['wordUrl'] },
    ];

    const propsMock = {
      listId: 'listIdMock',
      wordsArray,
      listName: 'listNameMock',
      addWord: jest.fn(),
      deleteWord: jest.fn(),
      openFlashcard: jest.fn(),
      openExerciseWindow: jest.fn(),
    };

    beforeEach(() => {
      propsMock.addWord.mockReturnValue(() => {});
      propsMock.openExerciseWindow.mockReturnValue(() => {});
      propsMock.deleteWord.mockReturnValue(() => {});
      propsMock.openFlashcard.mockReturnValue(() => {});
      mountedListItem = listItem(propsMock);
    });

    it('curries the deleteWord prop with the listId', () => {
      const { listId } = propsMock;
      expect(deleteWord).toHaveBeenCalledWith(listId);
    });

    it('passes it the curried deleteWord to the WordItem component', () => {
      const wordItem = mountedListItem.find('WordItem');
      const { listId } = propsMock;
      const curried = deleteWord(listId);
      expect(wordItem.first().props().deleteWord).toEqual(curried);
    });

    it('curries the openFlashcard prop with the listId', () => {
      const { openFlashcard, listId } = propsMock;
      expect(openFlashcard).toHaveBeenCalledWith(listId);
    });

    it('passes it the curried openFlashcard to the WordItem component', () => {
      const wordItem = mountedListItem.find('WordItem');
      const { openFlashcard, listId } = propsMock;
      const curried = openFlashcard(listId);
      expect(wordItem.first().props().openFlashcard).toEqual(curried);
    });

  });
});
