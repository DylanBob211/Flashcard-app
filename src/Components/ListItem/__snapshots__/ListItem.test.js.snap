// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ListItem when wordsArray prop empty matches the snapshot 1`] = `
<div
  className="listItem_container"
  data-test="listItemContainer"
>
  <ListHeader
    deleteList={[MockFunction]}
    listName="listNameMock"
    openExerciseWindow={[Function]}
  />
  <ul
    className="listItem_wordList"
  />
  <WordForm
    addWord={[Function]}
  />
</div>
`;
